File: src/semp_workflow/modules/acl_pub_exc.py
"""ACL Publish Topic Exception module - acl_publish_exception.add, acl_publish_exception.delete."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient
from ..semp.helpers import enc
from .base import BaseModule

logger = logging.getLogger(__name__)


class AclPublishExceptionAdd(BaseModule):
    description = "Add a publish topic exception to an ACL profile. Skipped if the exception already exists."
    params = {
        "aclProfileName":                {"type": "string", "required": True,  "description": "Name of the ACL profile"},
        "publishTopicException":         {"type": "string", "required": True,  "description": "The topic for the exception (may include wildcards)"},
        "publishTopicExceptionSyntax":   {"type": "string", "required": False, "description": "Syntax of the topic", "default": "smf", "enum": ["smf", "mqtt"]},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        profile = args.get("aclProfileName", "")
        topic = args.get("publishTopicException", "")
        syntax = args.get("publishTopicExceptionSyntax", "smf")

        if not profile:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: aclProfileName")
        if not topic:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: publishTopicException")

        path = f"aclProfiles/{enc(profile)}/publishTopicExceptions/{enc(syntax)},{enc(topic)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking publish topic exception: {e}")

        if exists:
            return ActionResult(
                ResultStatus.SKIPPED,
                f"Publish topic exception '{topic}' already exists on ACL profile '{profile}'",
            )

        if dry_run:
            return ActionResult(
                ResultStatus.DRYRUN,
                f"Would add publish topic exception '{topic}' to ACL profile '{profile}'",
            )

        try:
            payload = {
                "aclProfileName": profile,
                "publishTopicException": topic,
                "publishTopicExceptionSyntax": syntax,
            }
            client.create(f"aclProfiles/{enc(profile)}/publishTopicExceptions", payload)
            return ActionResult(
                ResultStatus.OK,
                f"Publish topic exception '{topic}' added to ACL profile '{profile}'",
            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED,
                f"Failed to add publish topic exception '{topic}' to ACL profile '{profile}': {e}",
            )


class AclPublishExceptionDelete(BaseModule):
    description = "Remove a publish topic exception from an ACL profile. Skipped if the exception does not exist."
    params = {
        "aclProfileName":                {"type": "string", "required": True,  "description": "Name of the ACL profile"},
        "publishTopicException":         {"type": "string", "required": True,  "description": "The topic exception to remove"},
        "publishTopicExceptionSyntax":   {"type": "string", "required": False, "description": "Syntax of the topic", "default": "smf", "enum": ["smf", "mqtt"]},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        profile = args.get("aclProfileName", "")
        topic = args.get("publishTopicException", "")
        syntax = args.get("publishTopicExceptionSyntax", "smf")

        if not profile:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: aclProfileName")
        if not topic:
            return ActionResult(ResultStatus.FAILED, "Missing required arg: publishTopicException")

        path = f"aclProfiles/{enc(profile)}/publishTopicExceptions/{enc(syntax)},{enc(topic)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking publish topic exception: {e}")

        if not exists:
            return ActionResult(
                ResultStatus.SKIPPED,
                f"Publish topic exception '{topic}' does not exist on ACL profile '{profile}'",
            )

        if dry_run:
            return ActionResult(
                ResultStatus.DRYRUN,
                f"Would remove publish topic exception '{topic}' from ACL profile '{profile}'",
            )

        try:
            client.delete(path)
            return ActionResult(
                ResultStatus.OK,
                f"Publish topic exception '{topic}' removed from ACL profile '{profile}'",
            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED,
                f"Failed to remove publish topic exception '{topic}' from ACL profile '{profile}': {e}",
            )


MODULES = {
    "acl_publish_exception.add": AclPublishExceptionAdd,
    "acl_publish_exception.delete": AclPublishExceptionDelete,
}

File: internal/modules/acl_pub_exc.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("acl_publish_exception.add", aclPublishExceptionAdd{})
	register("acl_publish_exception.delete", aclPublishExceptionDelete{})
}

type aclPublishExceptionAdd struct{}

func (aclPublishExceptionAdd) Description() string {
	return "Add a publish topic exception to an ACL profile. Skipped if the exception already exists."
}

func (aclPublishExceptionAdd) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "aclProfileName", Type: "string", Required: true, Description: "Name of the ACL profile"},
		{Name: "publishTopicException", Type: "string", Required: true, Description: "The topic for the exception (may include wildcards)"},
		{Name: "publishTopicExceptionSyntax", Type: "string", Description: "Syntax of the topic", Default: "smf", Enum: []string{"smf", "mqtt"}},
	}
}

func (aclPublishExceptionAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	profile := argStr(args, "aclProfileName", "")
	topic := argStr(args, "publishTopicException", "")
	syntax := argStr(args, "publishTopicExceptionSyntax", "smf")

	if profile == "" {
		return failed("Missing required arg: aclProfileName")
	}
	if topic == "" {
		return failed("Missing required arg: publishTopicException")
	}

	path := fmt.Sprintf("aclProfiles/%s/publishTopicExceptions/%s,%s", semp.Enc(profile), semp.Enc(syntax), semp.Enc(topic))

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking publish topic exception: %s", err))
	}
	if exists {
		return skipped(fmt.Sprintf("Publish topic exception '%s' already exists on ACL profile '%s'", topic, profile))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would add publish topic exception '%s' to ACL profile '%s'", topic, profile))
	}

	payload := map[string]any{
		"aclProfileName":              profile,
		"publishTopicException":       topic,
		"publishTopicExceptionSyntax": syntax,
	}
	if _, err := c.Create(fmt.Sprintf("aclProfiles/%s/publishTopicExceptions", semp.Enc(profile)), payload); err != nil {
		return failed(fmt.Sprintf("Failed to add publish topic exception '%s' to ACL profile '%s': %s", topic, profile, err))
	}
	return ok(fmt.Sprintf("Publish topic exception '%s' added to ACL profile '%s'", topic, profile))
}

type aclPublishExceptionDelete struct{}

func (aclPublishExceptionDelete) Description() string {
	return "Remove a publish topic exception from an ACL profile. Skipped if the exception does not exist."
}

func (aclPublishExceptionDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "aclProfileName", Type: "string", Required: true, Description: "Name of the ACL profile"},
		{Name: "publishTopicException", Type: "string", Required: true, Description: "The topic exception to remove"},
		{Name: "publishTopicExceptionSyntax", Type: "string", Description: "Syntax of the topic", Default: "smf", Enum: []string{"smf", "mqtt"}},
	}
}

func (aclPublishExceptionDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	profile := argStr(args, "aclProfileName", "")
	topic := argStr(args, "publishTopicException", "")
	syntax := argStr(args, "publishTopicExceptionSyntax", "smf")

	if profile == "" {
		return failed("Missing required arg: aclProfileName")
	}
	if topic == "" {
		return failed("Missing required arg: publishTopicException")
	}

	path := fmt.Sprintf("aclProfiles/%s/publishTopicExceptions/%s,%s", semp.Enc(profile), semp.Enc(syntax), semp.Enc(topic))

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking publish topic exception: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("Publish topic exception '%s' does not exist on ACL profile '%s'", topic, profile))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would remove publish topic exception '%s' from ACL profile '%s'", topic, profile))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to remove publish topic exception '%s' from ACL profile '%s': %s", topic, profile, err))
	}
	return ok(fmt.Sprintf("Publish topic exception '%s' removed from ACL profile '%s'", topic, profile))
}
